csrc/ir/utils.cpp

```diff
@@ -1310,4 +1310,111 @@ MmaOpDetails getMmaOpDetails(
+  NVF_ERROR(
+      in_b->as<TensorView>()->getRootDomain().front()->isBroadcast(),
+      "Expected first dimension to be broadcasted for second operand.");
+
+  // ATen preserves the dtype of MmaOp inputs whereas MmaOp generates float
+  // outputs. To preserve numerical equivalence and precision, the output of
+  // ATen matmul should be the same as MmaOp out `eventually`.
+  // See https://github.com/NVIDIA/Fuser/pull/1874#discussion_r1516991574
+  // Supported cases:
+  //  1. MmaOp->out() and MmaOp->input() are the same dtype.
+  //  2. MmaOp->out() is followed by a CastOp() to the MmaOp->input() dtype.
+  // NOTE: Currently MmaOp only accepts Half and BFloat16 so case (1) does not
+  // occur.
+
+  NVF_ERROR(
+      in_a->as<TensorView>()->getDataType().value() == final_out_dtype,
+      "CastOp should cast to the same final datatype as the input datatype.");
+}
+
+bool matchMatmulCast(const UnaryOp* cast_op, Val*& mma_lhs, Val*& mma_rhs) {
+  if (auto* mma = dynamic_cast<MmaOp*>(cast_op->input(0)->definition())) {
+    MmaOpUtils::verifyMmaOpForEvaluation(
+        mma, cast_op->out()->getDataType().value());
+    mma_lhs = mma->inA();
+    mma_rhs = mma->inB();
+    return true;
+  }
+  return false;
+}
+
+bool matchMatmulBiasCast(
+    const UnaryOp* cast_op,
+    Val*& mma_lhs,
+    Val*& mma_rhs,
+    Val*& bias) {
+  if (auto* binary = dynamic_cast<BinaryOp*>(cast_op->input(0)->definition())) {
+    if (binary->getBinaryOpType() == BinaryOpType::Add) {
+      if (auto* mma = dynamic_cast<MmaOp*>(binary->input(0)->definition())) {
+        DataType final_out_dtype = cast_op->out()->getDataType().value();
+
+        // Verify assumptions for MmaOp hold. Assign the values to Mma operands.
+        MmaOpUtils::verifyMmaOpForEvaluation(mma, final_out_dtype);
+        mma_lhs = mma->inA();
+        mma_rhs = mma->inB();
+
+        Val* bcast_bias = binary->input(1);
+        // Bias is casted to fp32 and broadcasted from shape [M,] to [M, 1] in
+        // biasEpilogue.
+        NVF_ERROR(
+            bcast_bias->definition() != nullptr &&
+                bcast_bias->definition()->isA<BroadcastOp>(),
+            "Expected bias tensor to be broadcasted.");
+
+        NVF_ERROR(
+            bcast_bias->as<TensorView>()->getRootDomain().back()->isBroadcast(),
+            "Expected last dimension to be broadcasted for bias tensor.");
+
+        Val* upcast_bias =
+            bcast_bias->definition()->as<BroadcastOp>()->input(0);
+
+        // The bias tensor and matmul inputs should be of the same dtype.
+        // The bias tensor is upcasted to fp32.
+        if (auto* unary = dynamic_cast<UnaryOp*>(upcast_bias->definition())) {
+          if (unary->getUnaryOpType() == UnaryOpType::Cast) {
+            NVF_CHECK(
+                unary->input(0)->getDataType().value() == final_out_dtype,
+                "Expected the bias to be originally of the same type as the final output dtype.");
```

Comment:
Is an `if` any safer? Nothing prevents the input fusion from having a different cast type. When that happens, is it better to hard fail or bail out? 

A similar comment goes to line 1413.

Reply:
I am not clear. Are you saying we should not hard fail here?
At this point, we have discovered a bias and MmaOp so the bias must fullfill all the conditions we expect under the current matmul structure.

Reply:
If there is not a castOp preceding the broadcasted bias, that implies the bias and the MmaOp inputs are of different types which is not supported in at::addmm, so we should hard fail. That implies we have an unexpected case / matmul fusion assumptions have changed and these need to be updated.

Reply:
> Expected the bias to be originally of the same type as the final output dtype.

Let me ask this in a different way: when you say "expect" here, which of the following do you mean
1. If the bias is not of the same type, it's a programming error in nvFuser and we should fix that programming error.
2. If the bias is not of the same time, it's still a valid input fusion (e.g. can be constructed using fd.ops) and we should evaluate these ops in a different way (e.g. evaluate them one at a time). 
3. If the bias is not of the same time, it's still a valid input fusion but we don't have an alternative to evaluate them, so let's hard fail. 

I thought it was 2 but after a second thought it is likely 3.

Reply:
I mean 3. similar to the case in `matmul + cast`. There is no corresponding ATen call.

Reply:
I can rephrase the error message.